File: spacecrypt.js
/**
 * Adapted from https://pug.sh/spacecrypt/ by Adam Newbold https://twitter.com/nwbld
 * @author Loreto Parisi (loretoparisi at gmail dot com)
 */

function str2bin(str, spaceSeparatedOctets) {
    function zeroPad(num) {
        return "00000000".slice(String(num).length) + num;
    }

    return str.replace(/[\s\S]/g, function (str) {
        str = zeroPad(str.charCodeAt().toString(2));
        return !1 == spaceSeparatedOctets ? str : str + " "
    });
};


function bin2str(str) {
    // Removes the spaces from the binary string
    str = str.replace(/\s+/g, '');
    // Pretty (correct) print binary (add a space every 8 characters)
    str = str.match(/.{1,8}/g).join(" ");

    var newBinary = str.split(" ");
    var binaryCode = [];

    for (i = 0; i < newBinary.length; i++) {
        binaryCode.push(String.fromCharCode(parseInt(newBinary[i], 2)));
    }

    return binaryCode.join("");
}

function bin2hiddenHTML(str) {
    str = str.replace(' ', '&#8288;'); // Unicode Character 'WORD JOINER' (U+2060) 0xE2 0x81 0xA0
    str = str.replace('0', '&#8203;'); // Unicode Character 'ZERO WIDTH SPACE' (U+200B) 0xE2 0x80 0x8B
    sstr = str.replace('1', '&#8204;'); // Unicode Character 'ZERO WIDTH NON-JOINER' (U+200C) 0xE2 0x80 0x8C
    return str;
}

function hidden2binHTML(str) {
    str = str.replace('&#8288;', ' ');
    str = str.replace('&#8203;', '0');
    str = str.replace('&#8204;', '1');
    return str;
}

function bin2hidden(str) {
    str = str.replace(' ', "\xE2\x81\xA0"); // Unicode Character 'WORD JOINER' (U+2060) 0xE2 0x81 0xA0
    str = str.replace('0', "\xE2\x80\x8B"); // Unicode Character 'ZERO WIDTH SPACE' (U+200B) 0xE2 0x80 0x8B
    sstr = str.replace('1', "\xE2\x80\x8C"); // Unicode Character 'ZERO WIDTH NON-JOINER' (U+200C) 0xE2 0x80 0x8C
    return str;
}

function hidden2bin(str) {
    str = str.replace("\xE2\x81\xA0", ' ');
    str = str.replace("\xE2\x80\x8B", '0');
    str = str.replace("\xE2\x80\x8C", '1');
    return str;
}

// TEST
var public = 'Hello World';
var private = 'What a fuck';
// private text
var binary = str2bin(private);
// hidden text
var hidden = bin2hidden(binary);

var encoded = hidden+' '+public;
var decoded = bin2str(hidden2bin(encoded));

console.log("Public:",public);
console.log("Binary:",binary);
console.log("Encoded binary:",hidden);
console.log("Decoded binary:",hidden2bin(encoded));
console.log("Private:",decoded);

File: spacecrypt.py
#!/usr/bin/python -tt
# -*- coding: utf-8 -*-

# Adapted from https://pug.sh/spacecrypt/ by Adam Newbold https://twitter.com/nwbld
# @author M. Stella Tavella (mstella.tavella at gmail dot com)


def str2bin(s):
    return ' '.join(format(ord(i),'b').zfill(8) for i in s) #2

def hasNumbers(inputString):
    return any(char.isdigit() for char in inputString)

def text_from_bits(bits, encoding='utf-8', errors='surrogatepass'):
    n = int(bits, 2)
    return n.to_bytes((n.bit_length() + 7) // 8, 'big').decode(encoding, errors) or '\0'

def bin2str(s):
    # Removes the spaces from the binary string
    s = s.replace(" ", "")

    # Pretty (correct) print binary (add a space every 8 characters)
    s = ' '.join([s[i:i+8] for i in range(0, len(s), 8)]) 

    newBinary = s.split(" ")

    newBinary = [i for i in newBinary if hasNumbers(i)]
    newBinary = ''.join(newBinary)

    binaryCode = text_from_bits(newBinary)

    return ''.join(binaryCode)

def bin2hiddenHTML(s):
    s = s.replace(' ', '&#8288;') # Unicode Character 'WORD JOINER' (U+2060) 0xE2 0x81 0xA0
    s = s.replace('0', '&#8203;') # Unicode Character 'ZERO WIDTH SPACE' (U+200B) 0xE2 0x80 0x8B
    s = s.replace('1', '&#8204;') # Unicode Character 'ZERO WIDTH NON-JOINER' (U+200C) 0xE2 0x80 0x8C
    return s

def hidden2binHTML(s):
    s = s.replace('&#8288;', ' ')
    s = s.replace('&#8203;', '0')
    s = s.replace('&#8204;', '1')
    return s

def bin2hidden(s):
    s = s.replace(' ', "\xE2\x81\xA0") # Unicode Character 'WORD JOINER' (U+2060) 0xE2 0x81 0xA0
    s = s.replace('0', "\xE2\x80\x8B") # Unicode Character 'ZERO WIDTH SPACE' (U+200B) 0xE2 0x80 0x8B
    s = s.replace('1', "\xE2\x80\x8C") # Unicode Character 'ZERO WIDTH NON-JOINER' (U+200C) 0xE2 0x80 0x8C
    return s

def hidden2bin(s):
    s = s.replace("\xE2\x81\xA0", ' ')
    s = s.replace("\xE2\x80\x8B", '0')
    s = s.replace("\xE2\x80\x8C", '1')
    return s


if __name__ == "__main__":

    # TEST
    public = 'Hello World'
    private = 'What a fuck'

    # private text
    binary = str2bin(private)

    # hidden text
    hidden = bin2hidden(binary)

    encoded = hidden + ' ' + public

    decoded = bin2str(hidden2bin(encoded))

    print("Public:", public)
    print("Binary:", binary)
    print("Encoded binary:", hidden)
    print("Decoded binary:", hidden2bin(encoded))
    print("Private:", decoded)
